Update the Nix build

sellout opened the pull request:
This brings everything Nix-related up to date.[^1]

[^1]: The GHC update also affects all Stack builds and the Cabal, HLS, and Stack updates also affect VS Code users.

- Cabal 3.10.3.0 → 3.14.2.0
- GHC 9.6.5 → 9.6.7
- haskell-language-server 2.8.0.0 → 2.10.0.0
- hpack 0.35.2 → 0.38.0
- Ormolu 0.7.2.0 → 0.7.4.0
- Stack 2.15.5 → 3.5.1

It also shrinks the closure size (mostly by eliminating a number of Nixpkgs inputs), adopts the nix-systems pattern to
allow downstream control over the included systems, and makes a number of other minor changes.

sellout committed:
Adopt nix-systems pattern

Somewhat described at https://github.com/nix-systems/nix-systems

This is useful for Haskell projects, because commands like

```bash
nix flake show --allow-import-from-derivation
```

can often fail, complaining that it can’t find some other architecture
to build on. With nix-systems, you can append
`--override-input systems github:nix-systems/x86_64-linux` to the
command to ensure that x86_64-linux is the only platform in play.

This also adds aarch64-linux to the matrix, since that is one of the
“default” systems in flake-utils.

sellout committed:
Update link to Stack docs

sellout committed:
Enable Hoogle for local packages

It is not triggering a(nother) rebuild of GHC.

sellout committed:
Remove obsolete Nix workaround

This workaround required another local workaround. If you had a line
like `LD_LIBRARY_PATH=` in your .envrc, it can be removed now.

sellout committed:
Update Cabal & HLS versions

This ensures we’re using the versions in Nixpkgs.

- cabal 3.10.3.0 → 3.14.2.0
- haskell-language-server 2.8.0.0 → 2.10.0.0

sellout committed:
Set up hpack as a haskell.nix tool

This allows us to install a different version than the one included in
Nixpkgs (although we don’t currently do that). It also makes some
syntactic changes to other tools for consistency.

sellout committed:
Don’t sandbox `unison-cli:test:cli-tests` in Nix

On macOS, it requires access to `security`, which is outside the
sandbox, so don’t sandbox it when running on Darwin.

This is made possible by setting `sandbox` to `relaxed` in the flake,
which allows using `__noChroot` to mark derivations as unsandboxed. This
makes things less strict on Linux (where the sandbox is on by default),
but more strict on Darwin (where the sandbox is off by default).

sellout committed:
Update Nix flake inputs

sellout committed:
Unify Nixpkgs inputs

This reduces the closure size a bit, and puts us on a consistent Nixpkgs
version.

This forces an upgrade of Stack from 2.15.7 to 3.5.1, and
correspondingly an upgrade of hpack from 0.35.2 to 0.38.0.

sellout committed:
Update Ormolu version

This ensures we’re using the version in Nixpkgs. It also reformats
everything.

- ormolu 0.7.2.0 → 0.7.4.0

sellout committed:
Update GHC to 9.6.7

Updating Nixpkgs resulted in a pile of “GHC 9.6.5 is out of date”
warnings, so this brings us to the end of the 9.6 line.

GHC is managed by the Stack resolver, so we can upgrade to GHC 9.8.4 or
9.10.2 without modifying the flake.

ceedubs commented:
From the Nix side this looks reasonable to me, but I vaguely remember there being some discussions in the past about specific Stack versions and what we should be on. I think that the main thing is just to keep people on roughly the same Stack versions? Anyway @aryairani, @mitchellwrosen, and @dolio (whom I don't think cares about the Stack version but probably does GHC) should probably be the ones to weigh in on that.

sellout commented:
> I vaguely remember there being some discussions in the past about specific Stack versions and what we should be on. I think that the main thing is just to keep people on roughly the same Stack versions?

Right. I think nix/versions.nix (which imports .vscode/settings.json) is the current best effort for that. That file could use a bit more documentation, but it’s
- checked/enforced by the Nix build,
- enforced(?) by the VS Code Haskell plugin,
- checked by scripts/check-formatting, and
- linked from [development.markdown](https://github.com/unisonweb/unison/blob/trunk/development.markdown#development-dependencies) for contributors whose tooling doesn’t handle it automatically[^1].

[^1]: And even the automatic stuff is not as great as it could be – e.g., VS Code lets you specify the HLS version, but not the version of Ormolu that HLS will use; and it lets you specify the Stack version, but not the version of hpack[^2] that Stack will use.

[^2]: At least with hpack, if we stay on a fairly recent version, we prevent accidental reversion, but we’re not so lucky with Ormolu.

IMO, places that set versions outside of nix/versions.nix[^3] are bugs to be fixed. E.g., we could templatize files that mention versions (e.g., GitHub workflows, .vscode/settings.json, check-formatting), and regenerate them during some part of the build.

[^3]: This could also become like versions.yaml or something, if we ever use it for generating files, to avoid requiring Nix for that process.

That said, I’m happy to get pushback on the specific versions (some are easier to convince Nix to change than others) and on the overall method for keeping versions in sync.

sellout committed:
Merge branch 'trunk' into update-nix

ceedubs commented:
This derivation builds fine for me, but when I try to `stack build` in the resulting shell I get a bunch of errors like:

```
       /home/unison/code/unison/codebase2/util-serialization/U/Util/Serialization.hs:58:21: error: [GHC-83865]                                                                                                      
           • Couldn't match expected type ‘bytestring-0.11.5.4:Data.ByteString.Internal.Type.ByteString’                                                                                                            
                         with actual type ‘ByteString’                                                                                                                                                              
             NB: ‘bytestring-0.11.5.4:Data.ByteString.Internal.Type.ByteString’                                                                                                                                     
                   is defined in ‘Data.ByteString.Internal.Type’                                                                                                                                                    
                       in package ‘bytestring-0.11.5.4’                                                                                                                                                             
                 ‘ByteString’                                                                                                                                                                                       
                   is defined in ‘Data.ByteString.Internal.Type’                                                                                                                                                    
                       in package ‘bytestring-0.11.5.4’                                                                                                                                                             
           • In the second argument of ‘runGetS’, namely ‘bytes’                                                                                                                                                    
             In the expression: runGetS getA bytes                                                                                                                                                                  
             In the expression:                                                                                                                                                                                     
               case runGetS getA bytes of                                                                 
                 Left _ -> Nothing                                                                                                                                                                                  
                 Right a -> Just a                                                                                                                                                                                  
          |                                                                                                                                                                                                         
       58 |   case runGetS getA bytes of Left _ -> Nothing; Right a -> Just a                                                                                                                                       
          |                     ^^^^^
```

No amount of deleting stack files seems to help with this.

sellout committed:
Don’t use the IOG cache at all

We had conditionally enabled it on Linux, but it just made the builds
inconsistent.

sellout committed:
Merge branch 'trunk' into update-nix

sellout commented:
This is superseded by #6033 and/or #6046.
